File: proto/edge/v1/ack_bounds_corpus_test.go
/*
 * Copyright 2026 Carver Automation Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package edgev1_test

import (
	"bytes"
	"fmt"
	"strings"
	"testing"

	"github.com/carverauto/serviceradar/go/pkg/edge/edgerecord"
	edgev1 "github.com/carverauto/serviceradar/proto/edge/v1"
	"google.golang.org/protobuf/proto"
)

func TestAckAdmissionCorpus(t *testing.T) {
	session := edgerecord.Session{SpoolID: uuidv7(1), Nonce: uuidv7(2), HighestSent: 3, SentEvents: map[uint64][]byte{1: uuidv7(11), 2: uuidv7(12), 3: uuidv7(13)}}
	goldenText(t, "ack_bound_session.txt", fmt.Sprintf("%x %x %x %x %x\n", session.SpoolID, session.Nonce, session.SentEvents[1], session.SentEvents[2], session.SentEvents[3]))
	var manifest strings.Builder
	emit := func(name, group string, accepted bool, raw []byte, wire, count, canonical int, s edgerecord.Session) {
		t.Helper()
		_, err := edgerecord.DecodeAck(raw, s, wire, count, canonical)
		if (err == nil) != accepted {
			t.Fatalf("%s accepted=%v got %v", name, accepted, err)
		}
		filename := "ack_bound_" + name + ".bin"
		goldenBytes(t, filename, raw)
		sent := 1
		if len(s.SentEvents) == 0 {
			sent = 0
		}
		fmt.Fprintf(&manifest, "%s %s %d %d %d %d %d %d %t\n", filename, group, wire, count, canonical, s.HighestSent, s.ResolvedThrough, sent, accepted)
	}
	marshal := func(a *edgev1.EdgeDeliveryAckV1) []byte {
		raw, err := proto.Marshal(a)
		if err != nil {
			t.Fatal(err)
		}
		return raw
	}
	retry := func(n int, code string) *edgev1.EdgeDeliveryAckV1 {
		a := &edgev1.EdgeDeliveryAckV1{SpoolId: session.SpoolID, SessionNonce: session.Nonce}
		for i := 0; i < n; i++ {
			a.Dispositions = append(a.Dispositions, &edgev1.EdgeRecordDisposition{Sequence: uint64(i + 1), Kind: edgev1.EdgeRecordDispositionKind_EDGE_RECORD_DISPOSITION_KIND_REJECTED_RETRYABLE, RejectionCode: code})
		}
		return a
	}
	for _, size := range []int{128, 129} {
		raw := marshal(retry(1, "A"))
		gap := size - len(raw)
		if gap%2 == 1 {
			raw = append(raw, 0x10, 0x80, 0)
		}
		for len(raw) < size {
			raw = append(raw, 0x10, 0)
		}
		if len(raw) != size {
			t.Fatal("raw padding")
		}
		emit(fmt.Sprintf("raw_%d", size), "raw", size == 128, raw, 128, 10, 100, session)
	}
	emit("raw_predecode", "raw", false, bytes.Repeat([]byte{255}, 129), 128, 10, 100, session)
	for _, n := range []int{2, 3} {
		emit(fmt.Sprintf("count_%d", n), "count", n == 2, marshal(retry(n, "A")), 1024, 2, 512, session)
	}
	for _, size := range []int{65, 66} {
		raw := marshal(retry(1, strings.Repeat("A", size-44)))
		if len(raw) != size {
			t.Fatalf("canonical size %d got %d", size, len(raw))
		}
		emit(fmt.Sprintf("canonical_%d", size), "canonical", size == 65, raw, 1024, 10, 65, session)
	}
	for _, tc := range []struct {
		name, code string
		accepted   bool
	}{
		{"code_empty", "", false}, {"code_min", "A", true}, {"code_cap", strings.Repeat("A", 64), true}, {"code_over", strings.Repeat("A", 65), false},
		{"code_lower", "Aa", false}, {"code_punctuation", "A-", false}, {"code_alphabet", "A_09Z", true},
	} {
		emit(tc.name, "code", tc.accepted, marshal(retry(1, tc.code)), 1024, 10, 512, session)
	}

	base := retry(1, "")
	base.ResolvedThroughSequence = 1
	base.Dispositions[0].Kind = edgev1.EdgeRecordDispositionKind_EDGE_RECORD_DISPOSITION_KIND_ACCEPTED_AUTHORITATIVE
	base.Dispositions[0].EventId = session.SentEvents[1]
	emit("accept_control", "semantic", true, marshal(base), 1024, 10, 512, session)
	for _, tc := range []struct {
		name     string
		accepted bool
		change   func(*edgev1.EdgeDeliveryAckV1)
	}{
		{"accept_audit", true, func(a *edgev1.EdgeDeliveryAckV1) {
			a.Dispositions[0].Kind = edgev1.EdgeRecordDispositionKind_EDGE_RECORD_DISPOSITION_KIND_ACCEPTED_AUDIT_ONLY
		}},
		{"accept_quarantine", true, func(a *edgev1.EdgeDeliveryAckV1) {
			a.Dispositions[0].Kind = edgev1.EdgeRecordDispositionKind_EDGE_RECORD_DISPOSITION_KIND_ACCEPTED_QUARANTINE
		}},
		{"permanent_absent_id", true, func(a *edgev1.EdgeDeliveryAckV1) {
			a.Dispositions[0].Kind = edgev1.EdgeRecordDispositionKind_EDGE_RECORD_DISPOSITION_KIND_REJECTED_PERMANENT
			a.Dispositions[0].EventId = nil
			a.Dispositions[0].RejectionCode = "A"
		}},
		{"spool", false, func(a *edgev1.EdgeDeliveryAckV1) { a.SpoolId = uuidv7(9) }},
		{"nonce", false, func(a *edgev1.EdgeDeliveryAckV1) { a.SessionNonce = uuidv7(9) }},
		{"window", false, func(a *edgev1.EdgeDeliveryAckV1) { a.ResolvedThroughSequence = 4 }},
		{"prefix", false, func(a *edgev1.EdgeDeliveryAckV1) { a.ResolvedThroughSequence = 2 }},
		{"sequence", false, func(a *edgev1.EdgeDeliveryAckV1) { a.Dispositions[0].Sequence = 2 }},
		{"kind_zero", false, func(a *edgev1.EdgeDeliveryAckV1) { a.Dispositions[0].Kind = 0 }},
		{"kind_unknown", false, func(a *edgev1.EdgeDeliveryAckV1) { a.Dispositions[0].Kind = 99 }},
		{"kind_negative", false, func(a *edgev1.EdgeDeliveryAckV1) { a.Dispositions[0].Kind = -1 }},
		{"event_absent", false, func(a *edgev1.EdgeDeliveryAckV1) { a.Dispositions[0].EventId = nil }},
		{"event_mismatch", false, func(a *edgev1.EdgeDeliveryAckV1) { a.Dispositions[0].EventId = uuidv7(9) }},
		{"event_short", false, func(a *edgev1.EdgeDeliveryAckV1) { a.Dispositions[0].EventId = []byte{1} }},
		{"accept_code", false, func(a *edgev1.EdgeDeliveryAckV1) { a.Dispositions[0].RejectionCode = "A" }},
		{"unknown_outer", false, func(a *edgev1.EdgeDeliveryAckV1) { a.ProtoReflect().SetUnknown([]byte{160, 6, 1}) }},
		{"unknown_nested", false, func(a *edgev1.EdgeDeliveryAckV1) { a.Dispositions[0].ProtoReflect().SetUnknown([]byte{160, 6, 1}) }},
	} {
		a := proto.Clone(base).(*edgev1.EdgeDeliveryAckV1)
		tc.change(a)
		emit(tc.name, "semantic", tc.accepted, marshal(a), 1024, 10, 512, session)
	}
	a := proto.Clone(base).(*edgev1.EdgeDeliveryAckV1)
	a.Dispositions = append(a.Dispositions, retry(2, "A").Dispositions[1])
	emit("retryable_tail", "semantic", true, marshal(a), 1024, 10, 512, session)
	a = retry(2, "A")
	a.ResolvedThroughSequence = 1
	a.Dispositions[1] = proto.Clone(base.Dispositions[0]).(*edgev1.EdgeRecordDisposition)
	a.Dispositions[1].Sequence = 2
	a.Dispositions[1].EventId = session.SentEvents[2]
	emit("resolve_after_retry", "semantic", false, marshal(a), 1024, 10, 512, session)
	missing := session
	missing.SentEvents = nil
	emit("missing_sent_event", "semantic", false, marshal(base), 1024, 10, 512, missing)
	short := session
	short.HighestSent = 0
	emit("past_highest_sent", "semantic", false, marshal(retry(1, "A")), 1024, 10, 512, short)
	prior := session
	prior.ResolvedThrough = 2
	emit("watermark_regression", "semantic", false, marshal(base), 1024, 10, 512, prior)
	high := session
	high.ResolvedThrough = ^uint64(0)
	high.HighestSent = ^uint64(0)
	a = retry(0, "A")
	a.ResolvedThroughSequence = high.ResolvedThrough
	emit("exhausted_empty", "semantic", true, marshal(a), 1024, 10, 512, high)
	a.Dispositions = retry(1, "A").Dispositions
	a.Dispositions[0].Sequence = 0
	emit("exhausted_nonempty", "semantic", false, marshal(a), 1024, 10, 512, high)
	emit("nonpositive_defaults", "policy", true, marshal(base), 0, -1, 0, session)
	goldenText(t, "ack_bounds_corpus.txt", manifest.String())
}

File: elixir/serviceradar_core/test/serviceradar/edge/ack_bounds_corpus_test.exs
defmodule ServiceRadar.Edge.AckBoundsCorpusTest do
  use ExUnit.Case, async: true

  alias ServiceRadar.Edge.AckValidate
  alias Serviceradar.Edge.V1.EdgeDeliveryAckV1

  @testdata Path.expand("../../../../../proto/edge/v1/testdata", __DIR__)
  @manifest Path.join(@testdata, "ack_bounds_corpus.txt")
  @external_resource @manifest

  test "complete raw ACK boundary agrees with the shared Go corpus" do
    rows = rows()

    for row <- rows do
      result = AckValidate.validate_bytes(fixture(row.file), session(row), row.limits)

      if row.accepted do
        assert {:ok, _ack} = result, "#{row.file}: #{inspect(result)}"
      else
        assert {:error, reason} = result, row.file
        refute reason in [:session, :limits], row.file
        refute reason in [{:wire, :not_ready}, {:wire, :systemic}], row.file
        refute match?({:schema_unavailable, _}, reason), row.file

        case row.group do
          "raw" -> assert reason == {:wire, :too_large}, row.file
          "count" -> assert reason == :count, row.file
          "canonical" -> assert reason == :canonical_bytes, row.file
          "code" -> assert reason == :rejection_code, row.file
          _ -> :ok
        end
      end
    end

    assert MapSet.new(rows, & &1.group) ==
             MapSet.new(~w(raw count canonical code semantic policy))

    expected = ~w(raw_128 raw_129 raw_predecode count_2 count_3 canonical_65 canonical_66
      code_empty code_min code_cap code_over code_lower code_punctuation code_alphabet
      accept_control accept_audit accept_quarantine permanent_absent_id spool nonce window prefix sequence
      kind_zero kind_unknown kind_negative event_absent event_mismatch event_short accept_code unknown_outer
      unknown_nested retryable_tail resolve_after_retry missing_sent_event past_highest_sent watermark_regression
      exhausted_empty exhausted_nonempty nonpositive_defaults)

    assert Enum.sort(Enum.map(rows, & &1.file)) ==
             Enum.sort(Enum.map(expected, &("ack_bound_" <> &1 <> ".bin")))
  end

  test "raw, count and canonical groups each isolate their own budget" do
    for row <- rows(),
        row.group in ["raw", "count", "canonical"],
        not String.contains?(row.file, "predecode") do
      raw = fixture(row.file)
      ack = EdgeDeliveryAckV1.decode(raw)
      count = length(ack.dispositions)
      canonical = byte_size(EdgeDeliveryAckV1.encode(ack))

      case row.group do
        "raw" ->
          assert count < row.limits.dispositions and canonical < row.limits.canonical_bytes
          assert byte_size(raw) == row.limits.raw_bytes + if(row.accepted, do: 0, else: 1)

        "count" ->
          assert byte_size(raw) < row.limits.raw_bytes and canonical < row.limits.canonical_bytes
          assert count == row.limits.dispositions + if(row.accepted, do: 0, else: 1)

        "canonical" ->
          assert byte_size(raw) < row.limits.raw_bytes and count < row.limits.dispositions
          assert canonical == row.limits.canonical_bytes + if(row.accepted, do: 0, else: 1)
      end
    end
  end

  test "code length, grammar and non-emptiness have independent controls" do
    for {name, length} <- [
          {"empty", 0},
          {"min", 1},
          {"cap", 64},
          {"over", 65},
          {"lower", 2},
          {"punctuation", 2}
        ] do
      ack = "ack_bound_code_#{name}.bin" |> fixture() |> EdgeDeliveryAckV1.decode()
      assert byte_size(hd(ack.dispositions).rejection_code) == length
    end
  end

  test "caller-shape faults and invalid limits are contained" do
    row = Enum.find(rows(), & &1.accepted)
    raw = fixture(row.file)

    for limits <- [nil, %{raw_bytes: 1.0}, %{raw_bytes: :infinity}, %{misspelled: 10}] do
      assert {:error, :limits} = AckValidate.validate_bytes(raw, session(row), limits)
    end

    for state <- [
          nil,
          %{},
          %{session(row) | resolved_through: -1},
          %{session(row) | highest_sent: 1.0}
        ] do
      assert {:error, :session} = AckValidate.validate_bytes(raw, state, row.limits)
    end

    assert {:error, {:wire, :systemic}} =
             AckValidate.validate_bytes(nil, session(row), row.limits)
  end

  defp session(row) do
    [spool, nonce, first, second, third] =
      "ack_bound_session.txt"
      |> fixture()
      |> String.split()
      |> Enum.map(&Base.decode16!(&1, case: :mixed))

    %{
      spool_id: spool,
      nonce: nonce,
      highest_sent: row.highest,
      resolved_through: row.resolved,
      sent_events: if(row.sent, do: %{1 => first, 2 => second, 3 => third}, else: %{})
    }
  end

  defp rows do
    @manifest
    |> File.read!()
    |> String.split("\n", trim: true)
    |> Enum.map(fn line ->
      [file, group, wire, count, canonical, highest, resolved, sent, accepted] =
        String.split(line)

      %{
        file: file,
        group: group,
        limits: %{
          raw_bytes: String.to_integer(wire),
          dispositions: String.to_integer(count),
          canonical_bytes: String.to_integer(canonical)
        },
        highest: String.to_integer(highest),
        resolved: String.to_integer(resolved),
        sent: sent == "1",
        accepted: accepted == "true"
      }
    end)
  end

  defp fixture(name), do: File.read!(Path.join(@testdata, name))
end
